server-utils/src/suspendable_stream.rs

```diff
@@ -0,0 +1,70 @@
+use std::time::{Duration, Instant};
+use tokio::timer::Delay;
+use std::io;
+use tokio::prelude::*;
+
+/// A wrapper type for tokio::Incoming that stops accepting new connections
+/// for a specified amount of time once `ulimit` is reached
+pub struct SuspendableStream<S> {
+	stream: S,
+	delay: Duration,
+	timeout: Option<Delay>,
+}
+
+impl<S, I> SuspendableStream<S>
+	where S: Stream<Item=I, Error=io::Error>
```

Comment:
I think you could skip the bounds here, and just add `new` function for any `S` (not even a `Stream`)